File: algorithms/algorithms-cplusplus/leetcode/Question_0010_Regular_Expression_Matching.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  bool isMatch(string s, string p) {
    if (p.empty()) {
      return s.empty();
    }
    vector<vector<bool> > dp(s.size() + 1, vector<bool>(p.size() + 1, false));
    dp[s.size()][p.size()] = true;
    for (int i = s.size(); i >= 0; i--) {
      for (int j = p.size() - 1; j >= 0; j--) {
        bool match = i < s.size() && (s[i] == p[j] || p[j] == '.');
        if (j + 1 < p.size() && p[j + 1] == '*') {
          dp[i][j] = dp[i][j + 2] || (match && dp[i + 1][j]);
        } else {
          dp[i][j] = match && dp[i + 1][j + 1];
        }
      }
    }
    return dp[0][0];
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0010_Regular_Expression_Matching.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0010_Regular_Expression_Matching {
    public boolean isMatch(String s, String p) {
        if (p == null) {
            return s == null;
        }
        boolean[][] dp = new boolean[s.length() + 1][p.length() + 1];
        dp[s.length()][p.length()] = true;
        for (int i = s.length(); i>= 0; i--) {
            for (int j = p.length() - 1; j >= 0; j--) {
                boolean match = i < s.length() && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.');
                if (j + 1 < p.length() && p.charAt(j + 1) == '*') {
                    dp[i][j] = dp[i][j + 2] || (match && dp[i + 1][j]);
                } else {
                    dp[i][j] = match && dp[i + 1][j + 1];
                }
            }
        }
        return dp[0][0];
    }

    public boolean isMatch_recursion(String s, String p) {
        if (s == null) {
            return p == null;
        }
        if (p.length() == 0) {
            return s.length() == 0;
        }
        return help(s, p, 0, 0);
    }

    private boolean help(String s, String p, int i, int j) {
        // 注意，使用j判断
        if (j >= p.length()) {
            return i >= s.length();
        }
        boolean match = i < s.length() && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.');
        if (j + 1 < p.length() && p.charAt(j + 1) == '*') {
            return (match && help(s, p, i + 1, j)) || help(s, p, i, j + 2);
        } else {
            return match && help(s, p, i + 1, j + 1);
        }
    }
}
